lib/inertia_rails/middleware.rb

```diff
@@ -3,41 +3,78 @@ class Middleware
     def initialize(app)
       @app = app
     end
-  
+
     def call(env)
+      @env = env
+
       status, headers, body = @app.call(env)
       request = ActionDispatch::Request.new(env)
 
       ::InertiaRails.reset!
-  
-      return [status, headers, body] unless env['HTTP_X_INERTIA'].present?
-  
-      return force_refresh(request) if stale?(env['REQUEST_METHOD'], env['HTTP_X_INERTIA_VERSION'])
-  
-      if is_redirect_status?(status) &&
-          is_non_get_redirectable_method?(env['REQUEST_METHOD'])
-        status = 303
-      end
-  
-      [status, headers, body]
-    end
-  
+
+      # Inertia errors are added to the session via inertia_redirect_to 
+      request.session.delete(:inertia_errors) unless keep_inertia_errors?(status)
+
+      status = 303 if inertia_non_post_redirect?(status)
+
+      return stale_inertia_get? ? force_refresh(request) : [status, headers, body]
+    end
+
     private
-  
-    def is_redirect_status?(status)
+
+    def keep_inertia_errors?(status)
+      redirect_status?(status) || stale_inertia_request?
+    end
```

Comment:
I _think_ this is the right logic for when to keep errors across requests, but i'm only 80% confident on that